plugins/catalyst/lib/Conversion/MQTOptToCatalystQuantum/MQTOptToCatalystQuantum.cpp

```diff
@@ -0,0 +1,1198 @@
+    // Extract operands
+    auto inQubitsValues = adaptor.getInQubits();
+    const auto posCtrlQubitsValues = adaptor.getPosCtrlInQubits();
+    const auto negCtrlQubitsValues = adaptor.getNegCtrlInQubits();
+
+    SmallVector<Value> inCtrlQubits(posCtrlQubitsValues.begin(),
+                                    posCtrlQubitsValues.end());
+    inCtrlQubits.append(negCtrlQubitsValues.begin(), negCtrlQubitsValues.end());
+
+    // Output type setup
+    const Type qubitType =
+        catalyst::quantum::QubitType::get(rewriter.getContext());
+    const SmallVector<Type> qubitTypes(
+        inQubitsValues.size() + inCtrlQubits.size(), qubitType);
+    auto outQubitTypes = TypeRange(qubitTypes);
+
+    // Peres = Toffoli(q2,q1,q0) CNOT(q2,q1)
+    auto ccnot1 = rewriter.create<catalyst::quantum::CustomOp>(
+        op.getLoc(), outQubitTypes, TypeRange{}, ValueRange{}, inQubitsValues,
+        "Toffoli", false, inCtrlQubits, ValueRange{});
+
+    auto cnot2 = rewriter.create<catalyst::quantum::CustomOp>(
+        op.getLoc(), outQubitTypes, TypeRange{}, ValueRange{},
+        ccnot1.getResults(), "CNOT", false, inCtrlQubits, ValueRange{});
+
+    rewriter.replaceOp(op, cnot2.getResults());
+    return success();
+  }
+};
+
+template <>
+struct ConvertMQTOptSimpleGate<opt::PeresdgOp> final
+    : OpConversionPattern<opt::PeresdgOp> {
+  using OpConversionPattern::OpConversionPattern;
+
+  LogicalResult
+  matchAndRewrite(opt::PeresdgOp op, opt::PeresdgOp::Adaptor adaptor,
+                  ConversionPatternRewriter& rewriter) const override {
+    // Extract operands
+    auto inQubitsValues = adaptor.getInQubits();
+    const auto posCtrlQubitsValues = adaptor.getPosCtrlInQubits();
+    const auto negCtrlQubitsValues = adaptor.getNegCtrlInQubits();
+
+    SmallVector<Value> inCtrlQubits(posCtrlQubitsValues.begin(),
+                                    posCtrlQubitsValues.end());
+    inCtrlQubits.append(negCtrlQubitsValues.begin(), negCtrlQubitsValues.end());
+
+    // Output type setup
+    const Type qubitType =
+        catalyst::quantum::QubitType::get(rewriter.getContext());
+    const SmallVector<Type> qubitTypes(
+        inQubitsValues.size() + inCtrlQubits.size(), qubitType);
+    auto outQubitTypes = TypeRange(qubitTypes);
+
+    // Peresdg = CNOT(q2,q1) Toffoli(q2,q1,q0)
+    auto cnot1 = rewriter.create<catalyst::quantum::CustomOp>(
+        op.getLoc(), outQubitTypes, TypeRange{}, ValueRange{}, inQubitsValues,
+        "CNOT", false, inCtrlQubits, ValueRange{});
+
+    auto ccnot2 = rewriter.create<catalyst::quantum::CustomOp>(
+        op.getLoc(), outQubitTypes, TypeRange{}, ValueRange{},
+        cnot1.getResults(), "Toffoli", false, inCtrlQubits, ValueRange{});
+
+    rewriter.replaceOp(op, ccnot2.getResults());
+    return success();
+  }
+};
+
+// -- GPhaseOp (gphase)
+template <>
+StringRef ConvertMQTOptSimpleGate<opt::GPhaseOp>::getGateName(
+    [[maybe_unused]] std::size_t numControls) {
+  return "gphase";
+}
```

Comment:
I am not 100% sure this is correct. If I remember correctly, Catalyst has its separate operation for the global phase. I might be mistaken, though.